k8s-friendly SSLEngine provider (simplified)

ignasi35 opened the pull request:
References #29146

Taking a fresh look at #29061

ignasi35 committed:
Refactor SSLEngineProviders to extract duplicate code

ignasi35 committed:
Code review. Cleanup, focus on cache. Removed duplication

ignasi35 committed:
Test extracted code

 - SecureRandomFactory
 - reader for X509 and peer verifier
 - JksManagersProviderSpec

ignasi35 committed:
Test PemManagersProvider and caching on the SSLFactory

ignasi35 committed:
Adopt akka-pki for PEM-based provision

ignasi35 committed:
Introduce CloudNativeSSLEngineProvider

ignasi35 committed:
Adds TlsMagic as (almost) copy/pasted from original

ignasi35 committed:
Start the refactor simplification

ignasi35 commented:
This takes #29061 a step back to simplify things a bit: (see 37d4e7e5cf9) 

 - remove the new `tcp.ssl.ConfigSSLEngineProvider`
 - undeprecate the old `tcp.ConfigSSLEngineProvider`
 - remove the `SSLFactory` class introduced in #29061 (too many things going on, too much indirection)
 - removed some no-longer-necessary abstractions (e.g. `ManagersProviders`)

Then, I've extracted a few bits and pieces from `AkkaTLSMagicSSLEngineProvider` to separate classes.

Finally, I've introduced `SSLEngineConfig` (related to the discussion in https://github.com/raboof/akka/pull/2#discussion_r431873610) and used it in both `AkkaTLSMagicSSLEngineProvider` and the old `ConfigSSLEngineProvider`.

The most important detail on that commit (37d4e7e5cf9) is that now there are only two `EngineProvider` implementations and only one of them support key rotation.

ignasi35 commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ConfigSSLEngineProvider.scala:
```diff
@@ -42,14 +42,18 @@ class ConfigSSLEngineProvider(protected val config: Config, protected val log: M
       system.settings.config.getConfig("akka.remote.artery.ssl.config-ssl-engine"),
       Logging.withMarker(system, classOf[ConfigSSLEngineProvider].getName))
 
+  val sslEngineConfig = new SSLEngineConfig(config)
```
I'm tempted to make this one `protected` and deprecate all five fields below.

raboof replied:
I'd like to keep this one private for now, so we can evolve it more freely later.

If we make it public now it means we also need to 'get it right' which seems ambitious.

ignasi35 committed:
Complete the key rotation test

ignasi35 committed:
Introduce akka-remote/ssl test resources (scripts, resources and tests)

ignasi35 committed:
PR comments

ignasi35 commented:
@raboof I've renamed the new engine provider to `RotatingKeysSSLEngine` and introduced the `RotatingKeysSSLEngineSpec` to test the `SSLContext` caching.

I was getting a bit lost wit so many, and varied, certificates so I've done a thorough clean up and documentation. I've also added a `TlsResourcesSpec` and a set of scripts to rebuild the resources (p12, pem, cert...) safely. The rule of thumb is: tests in `TlsResourcesSpec` are canary, if there's a failure in `TlsResourcesSpec` it's very likely things will break elsewhere.

I've rebased on `master` and squashed the history a bit but still kept the first half of the work (with the `CloudNative` impl and `SSLFactory`, etc...). We can squash more later.

ignasi35 commented on akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/RotatingKeysSSLEngineSpec.scala:
```diff
@@ -0,0 +1,152 @@
+      val after = remoteSysB.sslContextRef.get()
+      before shouldNot be(after)
+
+    }
```
Note: this tests only asserts a new `SSLContext` is built. It doesn't test keys on the FS are re-read yet.

ignasi35 committed:
Refactor rotating spec

ignasi35 committed:
Assert no new engines created

ignasi35 committed:
Tests existing connections are reused after cache expiration

ignasi35 committed:
Reviews certificate/key set for tests

ignasi35 committed:
typo

ignasi35 committed:
Run blocking IO in `blocking{}

ignasi35 committed:
Partial commit

ignasi35 committed:
WIP: an attempt to do a file-swatp test

raboof committed:
Remove unused parameter

to fix compilation

raboof committed:
Shut down systems before removing temp directory

raboof committed:
Fix compilation on Scala 2.13

raboof committed:
scalafmt

raboof committed:
Fix X509ReadersSpec

raboof committed:
Add some docs and Internal/MayChange markers

raboof committed:
fits on one line now

raboof committed:
Add comment on the status of to rotating provider

raboof committed:
Check temporary directory is ready for business

ignasi35 committed:
Cleanup unnecessary checks

ignasi35 committed:
Fix deletion/recreation of the temp folder

ignasi35 committed:
Cleanup TODOs

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -1176,6 +1175,66 @@ akka {
+          #
+          # Setting a value here may require you to supply the appropriate cipher
+          # suite (see enabled-algorithms section)
+          random-number-generator = ${akka.remote.artery.ssl.config-ssl-engine.random-number-generator}
```
Is it for our own convenience or for the users convenience that we pick values from `config-ssl-engine` here?
I'm worried that it's surprising that changing config-ssl-engine also changes rotating-keys-engine.

ignasi35 replied:
It's for our convenience. 

The settings reused are unrelated to the implementation and are actually an indicator of code and settings that should be extracted and reused across `SSLEngineProvider` implementations.

patriknw replied:
Then I'd prefer that we copy the value here to avoid surprises for users that one influence the other.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/X509Readers.scala:
```diff
@@ -0,0 +1,40 @@
+    val alternates = Option(cert.getSubjectAlternativeNames).map(_.asScala).getOrElse(Nil).collect {
+      // See the javadocs for what this list contains, first element should be an integer,
+      // if that integer is 2, then the second element is a String containing the DNS name.
+      case list if list.get(0) == 2 =>
```
can the list be empty? 
`getOrElse(Nil)` above?

ignasi35 replied:
Good catch. I'm not sure. This was copy pasted. I'll try to create a keyset with a certificate that has an empty list.

ignasi35 replied:
The return type of `getSubjectAlternativeNames` is a bit convoluted:

- if the certificate has no `SubjectAlternativeNames` (aka `SAN`), then `null` is returned. Then, we'd be running `Nil.collect(...)` so we'd collect over an empty list of `SAN`.
- else, a `Collection<List<?>>` is returned where each item of the `Collection<>` is a `SAN` represented as an heterogeneous list where `list.get(0)` is an integer identifying the SAN type (`DNS`, `EMAIL`, ...) and `list.get(1)` is the value. By contract of `getSubjectAlternativeNames` doing `list.get(0)` and `list.get(1)` is safe.

patriknw replied:
Thanks for clarifying. I missed that it was a collection of a collection

ignasi35 commented on akka-remote/src/test/scala/akka/remote/artery/tcp/TlsTcpSpec.scala:
```diff
@@ -53,6 +49,17 @@ class TlsTcpWithCrappyRSAWithMD5OnlyHereToMakeSureThingsWorkSpec
+         ca-cert-file = ${getClass.getClassLoader.getResource("ssl/exampleca.crt").getPath}
+       }
+    }
+    """))
```
A smoke test on `RotatingKeysSSLEnginProvider` (not the real deal).

ignasi35 commented on akka-remote/src/test/scala/akka/remote/artery/tcp/TlsTcpSpec.scala:
```diff
@@ -122,43 +123,6 @@ abstract class TlsTcpSpec(config: Config)
-        // Another reproducer of https://doc.akka.io/docs/akka/current/security/2018-08-29-aes-rng.html
-        // with the broken implementation the compressed size was <5k
-        compressed.size should be > randomData.length
-      }
```
Moved to its own Spec.

ignasi35 commented on akka-remote/src/test/scala/akka/remote/artery/tcp/TlsTcpSpec.scala:
```diff
@@ -257,17 +221,21 @@ class TlsTcpWithActorSystemSetupSpec extends ArteryMultiNodeSpec(TlsTcpSpec.conf
+      override def verifyClientSession(hostname: String, session: SSLSession): Option[Throwable] =
+        delegate.verifyClientSession(hostname, session)
+      override def verifyServerSession(hostname: String, session: SSLSession): Option[Throwable] =
+        delegate.verifyServerSession(hostname, session)
```
Prefer composition for future replacement of the type of `delegate`.

ignasi35 commented on akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/CipherSuiteSupportCheck.scala:
```diff
@@ -0,0 +1,68 @@
+   * Given a `configPath` runs a set of validations over the default protocols,
+   * cipher suites, PRNG algorithm/provider, etc...
+   */
+  def isSupported(system: ActorSystem, configPath: String): Try[Unit] = Try {
```
Extracted from `TlsTcpSpec`

ignasi35 commented on akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/TlsResourcesSpec.scala:
```diff
@@ -0,0 +1,94 @@
+/**
+ *
+ */
+class TlsResourcesSpec extends AnyWordSpec with Matchers {
```
This test ensures the files under `src/test/resources/ssl` follow certain properties (e.g. key algorythm, SAN values, key usage,...)

ignasi35 reviewed (commented):
Introduces `RotatingKeysSSLEngineProvider` (and supporting classes) under `akka.remote.artery.tcp.ssl`.

All new classes on `akka.remote.artery.tcp.ssl` have their own tests too.

ignasi35 commented:
Despite adding some comments on `reference.conf` wrt key types and other format restrictions to use `rotating-keys-engine` there could be extra considerations to document such as:

 - rolling upgrade considerations (moving from `config-ssl-engine` to `rotating-keys-engine`)
 - certificate duration, certificate overlap, cache duration, etc...
 - others?

If we deem necessary to add those docs, I'd rather do it on separate PRs.

ignasi35 committed:
Simplifies docs for resources. Increases coverage

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ConfigSSLEngineProvider.scala:
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
```
yes, confirmed (it's the same)

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/SecureRandomFactory.scala:
```diff
@@ -0,0 +1,58 @@
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object SecureRandomFactory {
```
this is the same as before

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/PemManagersProvider.scala:
```diff
@@ -0,0 +1,91 @@
+  @InternalApi
+  private[ssl] def loadCertificate(filename: String): Certificate = blocking {
+    val bytes = Files.readAllBytes(new File(filename).toPath)
+    certFactory.generateCertificate(new ByteArrayInputStream(bytes))
```
Is there a reason for using Files and ByteArrayInputStream instead of `new BufferedInputStream(new FileInputStream(filename))`?

ignasi35 replied:
No, this is a leftover. I was doing an experiment and committted this by mistake.

ignasi35 replied:
Hmm, on second thought. I think I prefer the code as is:

1. the method `loadPrivateKey` above uses the same idiom (in that case because a `Array[Byte]`  is required, though)
2. the `Files.readAllBytes` grants the underlying `FileInputStream` is closed so we don't need to implement the `try/finally` here

It's a pity that `CertificateFactory` factory methods require `InpuStream` and don't provide a `Array[Byte]` overload. An instance of a `ByteArrayInputStream` is required but this is not a hotspot anyway.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/RotatingKeysSSLEngineProvider.scala:
```diff
@@ -0,0 +1,161 @@
+    contextRef.get() match {
+      case Some(CachedContext(_, expired)) if expired.isOverdue() =>
+        val context = constructContext()
+        contextRef.set(Some(CachedContext(context, SSLContextCacheTime.fromNow)))
```
maybe add a comment that it's fine to create new context in case of concurrent access

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/RotatingKeysSSLEngineProvider.scala:
```diff
@@ -0,0 +1,161 @@
+  private val rng: SecureRandom = SecureRandomFactory.createSecureRandom(SSLRandomNumberGenerator, log)
+
+  // handle caching
+  private val contextRef = new AtomicReference[Option[CachedContext]](None)
```
could be volatile var since we don't use compare and swap
```
@volatile private var contextRef: Option[CachedContext] = None
```

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/SSLEngineConfig.scala:
```diff
@@ -0,0 +1,41 @@
+    if (config.hasPath("ssl-context-cache-ttl"))
+      config.getDuration("ssl-context-cache-ttl").toMillis.millis
+    else
+      1024.days // a lot of days (not Inf, because `Inf` is not a FiniteDuration
```
ok, that will only be the case when read from `config-ssl-engine`, and then not used

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/RotatingKeysSSLEngineProvider.scala:
```diff
@@ -0,0 +1,161 @@
+ * the configured certificate.
+ */
+@ApiMayChange
+final class RotatingKeysSSLEngineProvider(val config: Config, protected val log: MarkerLoggingAdapter)
```
good name

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/SessionVerifier.scala:
```diff
@@ -0,0 +1,62 @@
+ * Allows hooking in extra verification before finishing the SSL handshake.
+ */
+@ApiMayChange
+trait SessionVerifier {
```
I can't see that any of these `SessionVerifier` are exposed in public API so we can make them internal?

ignasi35 replied:
Good catch. This was public APi on a previous apprach I had. Privatizing...

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/X509Readers.scala:
```diff
@@ -0,0 +1,43 @@
+ * INTERNAL API
+ */
+@InternalApi
+object X509Readers {
```
`private[akka]`

ignasi35 replied:
I've been using the more strict `private[tcp]` or even `private[ssl]` in some places. Should I fallback to `private[akka]` elsewhere too?

patriknw replied:
Doesn't matter much. I often run into more trouble than what it's worth when using more strict private packages. One exposing something else in method and then compiler warns about visibility.

patriknw commented on akka-remote/src/test/scala/akka/remote/artery/ArteryMultiNodeSpec.scala:
```diff
@@ -58,7 +58,7 @@ abstract class ArteryMultiNodeSpec(config: Config)
     val config =
       extraConfig.fold(localSystem.settings.config)(str =>
-        ConfigFactory.parseString(str).withFallback(localSystem.settings.config))
+        ConfigFactory.parseString(str).resolve().withFallback(localSystem.settings.config))
```
is this still needed if we remove the `${...config-ssl-engine}` from reference conf, or what was the reason?

ignasi35 replied:
Tests in `RotatingKeysSSLEngineProviderSpec` were also using substitution. I've rewritten them to avoid it.

patriknw commented on akka-remote/src/test/scala/akka/remote/artery/ArterySpecSupport.scala:
```diff
@@ -32,6 +32,9 @@ object ArterySpecSupport {
   // TLS only used if transport=tls-tcp, which can be set from specific tests or
   // System properties (e.g. jenkins job)
+  // TODO: randomly return a Config using ConfigSSLEngineProvider or
+  //  CloudNativeSSLEngineProvider and, eventually, run tests twice
```
old name CloudNativeSSLEngineProvider

patriknw reviewed (commented):
Looking great, seems to be ready for merge very soon. Nice that you ended up with this simplified approach.

ignasi35 committed:
Some private markers

ignasi35 committed:
Don't reuse settings

ignasi35 committed:
Some more private markers

ignasi35 committed:
Simplify config, avoid resolve()

ignasi35 committed:
Comment typos

ignasi35 committed:
Use volatile var and comment about multple instances

ignasi35 commented:
Thanks @patriknw for all the comments. I've addressed all discussion points.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/RotatingKeysSSLEngineProvider.scala:
```diff
@@ -66,21 +66,26 @@ final class RotatingKeysSSLEngineProvider(val config: Config, protected val log:
 
   // handle caching
-  private val contextRef = new AtomicReference[Option[CachedContext]](None)
+  @volatile private var contextRef: Option[CachedContext] = None
```
rename to `context` or `cachedContext`?

patriknw reviewed (approved):
LGTM

ignasi35 committed:
Rename field

jroper commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -1176,6 +1175,66 @@ akka {
+
+          # There are two options, and the default SecureRandom is recommended:
+          # "" or "SecureRandom" => (default)
+          # "SHA1PRNG" => Can be slow because of blocking issues on Linux
```
Not sure why this is limited to just default or `SHA1PRNG`, I would have thought any algorithm that any available JSSE provider supports should be allowed - not that anyone would ever do it, but if you had a custom hardware random generator that you wanted to use, it would have an algorithm name, and then you would get it by passing that name to `SecureRandom.getInstance()`. Limiting it to just the default or `SHA1PRNG` seems arbitrary. I would both change the behavior of how this is supported to allow any value that just gets passed to `SecureRandom.getInstance`, and then like to [here](https://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#SecureRandom) for a list and discussion of valid values.

ignasi35 replied:
Good point. All the PRNG setup and related code `SecureRandomFactory` is provided as it already was (only extracted into a separate class).

We can make improvements on a separate PR.

raboof replied:
added #29188

jroper commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -1176,6 +1175,66 @@ akka {
+          enabled-algorithms = ["TLS_DHE_RSA_WITH_AES_128_GCM_SHA256",
+             "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
+             "TLS_DHE_RSA_WITH_AES_256_GCM_SHA384",
+             "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"]
```
Order here matters, it specifies the preference. Since typically all nodes will have the same configuration, this means `TLS_DHE_RSA_WITH_AES_128_GCM_SHA256` will always be selected, and the other three will never be considered. Is that the best out of the 4? I would have suggested `TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384` as the top one, since `ECDHE` is considered more secure than `DHE` (there are known ways to approximate factoring primes that lead to weaknesses in `DHE`, and new ways that make it weaker are always coming up), whereas there's no equivalent known weaknesses in eliptic curve cryptography), and `SHA384` is faster on 64 bit hardware than `SHA256`.

I would actually recommend only having one algorithm here too - there's just no reason to have more than one since both the client and server will share the same configuration, unless you're switching defaults, and then in one release you would add the new default as the first option and the old as the second, and then in a subsequent release, you would drop the second.

jroper commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -1176,6 +1175,66 @@ akka {
+             "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"]
+
+          # Protocol to use for SSL encryption, choose from:
+          # TLS 1.2 is available since JDK7, and default since JDK8:
```
Akka doesn't run on JDK7, doesn't it emit JDK8 byte code? Does it make sense to even mention JDK7 here?

ignasi35 replied:
This comment is triple-copied from providers above.

jroper commented on akka-remote/src/main/scala/akka/remote/artery/tcp/SecureRandomFactory.scala:
```diff
@@ -0,0 +1,58 @@
+
+  def createSecureRandom(randomNumberGenerator: String, log: MarkerLoggingAdapter): SecureRandom = {
+    val rng = randomNumberGenerator match {
+      case s @ (GeneratorSha1Prng | GeneratorNativePrng) =>
+        log.debug("SSL random number generator set to: {}", s)
+        // SHA1PRNG needs /dev/urandom to be the source on Linux to prevent problems with /dev/random blocking
+        // However, this also makes the seed source insecure as the seed is reused to avoid blocking (not a problem on FreeBSD).
+        SecureRandom.getInstance(s)
+
+      case "" | GeneratorJdkSecureRandom =>
+        log.debug("SSL random number generator set to [SecureRandom]")
+        new SecureRandom
+
+      case unknown =>
+        log.warning(
+          LogMarker.Security,
+          "Unknown SSL random number generator [{}] falling back to SecureRandom",
+          unknown)
+        new SecureRandom
```
I don't see any reason why we wouldn't just let users select whatever algorithm they want.

```suggestion
      case "" | GeneratorJdkSecureRandom =>
        log.debug("Using platform default SecureRandom algorithm for SSL")
        new SecureRandom
      case custom =>
        log.debug("Using {} SecureRandom algorithm for SSL", custom)
        SecureRandom.getInstance(custom)
```

If this suggestion is accepted, then the `GeneratorSha1Prng` and `NativePRNG` constants above sohuld also be deleted.

jroper commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/RotatingKeysSSLEngineProvider.scala:
```diff
@@ -0,0 +1,174 @@
+      val keyManagers: Array[KeyManager] = PemManagersProvider.buildKeyManagers(privateKey, cert, cacert)
+      val trustManagers: Array[TrustManager] = PemManagersProvider.buildTrustManagers(cacert)
+
+      val sessionVerifier = new PeerSubjectVerifier(cert)
```
Perhaps this is not needed in this PR, but it would be nice if the subject verification were configurable. For example, you might turn on/off whether SANs are included or if just the CN is used, and you might also want to specify a name or list of names that are allowed. An example where a list of names from configuration might be valuable is if you had different deployments with different roles, and for whatever reason deployment got their own certificate with their own CN, and you wanted all pods from all of these deployments to form a single cluster.

But, as it is, the peer subject verifier is probably what will be most convenient and useful for most users, so I think this PR is fine as is for now.

ignasi35 replied:
The thin line between configurable and too-much public API is what kept us busy when porting this code here. The introduction of the `SessionVerifier` trait is leading us into the direction of providing multiple verifiers and being able to configure an array of implementations on settings.

raboof replied:
created  #29189

jroper commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/X509Readers.scala:
```diff
@@ -0,0 +1,43 @@
+      new LdapName(cert.getSubjectX500Principal.getName).getRdns.asScala.collectFirst {
+        case attr if attr.getType.equalsIgnoreCase("CN") =>
+          attr.getValue.toString
+      }
```
May want to think about error handling here - failing/throwing is probably the right thing to do, but might want to wrap in an exception that gives more useful error info in what was trying to be achieved.

ignasi35 replied:
This is internal and only used in the `PeerSubjectVerifier` code. The only reason why `X509readers` is extracted is to unit test it.

I've been trying to cause a failure in the `X509readers` code with several certificates. All of the inputs I used in the tests are valid x.509 certificates. I've used different values for the CN in the `DName` and the `DNS` field in the SAN extension: no SAN, SAN without DNS, DName without CN, duplicate entries,... Even when there's no subject at all, `getAllSubjectNames()` returns the expected `Set()`. I haven't pushed this tests yet to not extend this PR even more.

The other possible failure I can think of could be caused if the file provided by the user is not a valid `X509Certificate`, in which case this code is unreachable.

If you have any other cases in mind, I think we can make the improvement on a separate PR.

ignasi35 replied:
pushed these tests in https://github.com/ignasi35/akka/commit/5da4fc6e4b514f6cce03d7014714b15052b98a1c

jroper commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/X509Readers.scala:
```diff
@@ -0,0 +1,43 @@
+      // See the javadocs of cert.getSubjectAlternativeNames for what this list contains,
+      // first element should be an integer, if that integer is 2, then the second element
+      // is a String containing the DNS name.
+      case list if list.get(0) == 2 =>
```
```suggestion
      case list if list.size() == 2 && list.get(0) == 2 =>
```

Perhaps better in this case, where there's a lot of things we don't know how to handle, to ignore things that don't match our expectations exactly, rather than throw a `NoSuchElementException`.

jroper commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/X509Readers.scala:
```diff
@@ -0,0 +1,43 @@
+        list.get(1) match {
+          case dnsName: String => dnsName
+          case other =>
+            throw new IllegalArgumentException(s"Expected a string, but got a ${other.getClass}")
```
```suggestion
            throw new IllegalArgumentException(s"Error reading Subject Alternative Name, expected dns name to be a String, but instead got a ${other.getClass}")
```

ignasi35 commented:
Hi @jroper, thanks for all the feedback. Kid just woke up so I'll need to go offline for a few hours. Will review the last comments this PM.

raboof committed:
skip alternatives that don't have the expected number of items

Co-authored-by: James Roper <james@jazzy.id.au>

raboof committed:
Improve exception text

Co-authored-by: James Roper <james@jazzy.id.au>

raboof commented on akka-pki/src/test/resources/README.md:
```diff
@@ -0,0 +1,3 @@
+# SSL/TLS-PEM resources
+
+TODO: add a script/tool to regenerate the resources here.
```
Perhaps remove this file if we're not going to introduce this script/tool yet?

ignasi35 replied:
Fair enough, I'll create a ticket instead.

ignasi35 replied:
Raised #29191

raboof committed:
scalafmt

ignasi35 committed:
Remove unnecessary file with a TODO

ignasi35 commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -1176,6 +1175,66 @@ akka {
+          # the JCE Unlimited Strength Jurisdiction Policy Files to use AES 256.
+          # More info here:
+          # https://www.oracle.com/java/technologies/javase-jce-all-downloads.html
+          enabled-algorithms = ["TLS_DHE_RSA_WITH_AES_128_GCM_SHA256",
+             "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
+             "TLS_DHE_RSA_WITH_AES_256_GCM_SHA384",
+             "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"]
```
```suggestion
          enabled-algorithms = ["TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"]
```

ignasi35 committed:
Prefer a single cipher suite

SJern commented:
Would you guys consider removing the `final` keyword from the `RotatingKeysSSLEngineProvider` class? We do not use k8s, but we do have the cert rotation requirement. It would be great if we could override some methods to customize how we read the certs and build BCFKS KeyStore.

patriknw commented:
@SJern The problem could be that it become difficult for us to change it if we have to take into consideration that some methods can be overridden in custom subclasses. Are there any methods or utilities there that we could instead move to the companion object and make public? Then you could implement your own `SSLEngineProvider` without extending `RotatingKeysSSLEngineProvider`?

SJern commented:
Hi @patriknw It would be nice if we could use the `buildKeyManagers` and `buildTrustManagers`, and also being able to choose our own type of keystore such as BCFKS, and our own keystore password.
`SessionVerifier.scala` would also be nice too.
In the future, if you guys decided to remove the final keyword, it would be nice if we could choose our own type of keystore such as BCFKS, and our own keystore password. Then we would just need to override the `readFiles` method.
Thanks!
